sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,752 @@
+  if ((rc = MQTTClient_publish(
+           mqtt_client,
+           reported_property_topic,
+           az_span_size(json_payload),
+           az_span_ptr(json_payload),
+           0,
+           0,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish reported property, return code %d\n", rc);
+    return rc;
+  }
+  return rc;
+}
+
+// Parse the desired temperature property from the incoming JSON
+static az_result parse_desired_temperature_property(az_span twin_span, double* parsed_value)
+{
+  az_json_parser jp;
+  az_json_token_member tm;
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, twin_span));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  while (az_succeeded(az_json_parser_parse_token_member(&jp, &tm)))
+  {
+    if (az_span_is_content_equal(desired_temp_property_name, tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm.token, parsed_value));
+    }
+
+    // else ignore token.
+  }
+
+  return AZ_OK;
+}
+
+// Callback for incoming MQTT messages
+static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+{
+  (void)context;
+
+  if (topicLen == 0)
+  {
+    // The length of the topic if there are one or more NULL characters embedded in topicName,
+    // otherwise topicLen is 0.
+    topicLen = (int)strlen(topicName);
+  }
+
+  printf("Topic: %s\n", topicName);
+
+  az_span topic_span = az_span_init((uint8_t*)topicName, topicLen);
+
+  // Parse the incoming message topic and check which feature it is for
+  az_iot_hub_client_twin_response twin_response;
+  az_iot_hub_client_method_request method_request;
+
+  if (az_succeeded(
+          az_iot_hub_client_twin_parse_received_topic(&client, topic_span, &twin_response)))
+  {
+    printf("Twin Message Arrived\n");
+
+    // Determine what type of incoming twin message this is. Print relevant data for the message.
+    switch (twin_response.response_type)
+    {
+      // A response from a twin GET publish message with the twin document as a payload.
+      case AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_GET:
+        printf("A twin GET response was received\n");
+        if (message->payloadlen)
+        {
+          printf("Payload:\n%.*s\n", message->payloadlen, (char*)message->payload);
+        }
+        break;
+      // An update to the desired properties with the properties as a JSON payload.
+      case AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES:
+        printf("A twin desired properties message was received\n");
+        printf("Payload:\n%.*s\n", message->payloadlen, (char*)message->payload);
+        az_span twin_span = az_span_init((uint8_t*)message->payload, (int32_t)message->payloadlen);
+
+        // Get the new temperature
+        double desired_temp;
+        parse_desired_temperature_property(twin_span, &desired_temp);
```

Comment:
`parse_desired_temperature_property` can end up returning success even if desired_temp is not found.  So even if we check its return value, we need to check that desired_temp was actually in the JSON or else code below will end up setting garbage.

Reply:
fixed